// typescript-docs-verifier/dist/src/SnippetCompiler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnippetCompiler = void 0;
const path = __importStar(require("path"));
const chalk_1 = __importDefault(require("chalk"));
const tsconfig = __importStar(require("tsconfig"));
const fsExtra = __importStar(require("fs-extra"));
const TSNode = __importStar(require("ts-node"));
const CodeBlockExtractor_1 = require("./CodeBlockExtractor");
const LocalImportSubstituter_1 = require("./LocalImportSubstituter");
class SnippetCompiler {
    constructor(workingDirectory, packageDefinition, project) {
        this.workingDirectory = workingDirectory;
        this.packageDefinition = packageDefinition;
        const configOptions = SnippetCompiler.loadTypeScriptConfig(packageDefinition.packageRoot, project);
        this.compilerConfig = {
            ...configOptions.config,
            transpileOnly: false,
        };
    }
    static loadTypeScriptConfig(packageRoot, project) {
        var _a;
        const fullProjectPath = path.join(packageRoot, project !== null && project !== void 0 ? project : "");
        const { base, dir } = path.parse(fullProjectPath);
        const typeScriptConfig = tsconfig.loadSync(dir, base);
        if ((_a = typeScriptConfig === null || typeScriptConfig === void 0 ? void 0 : typeScriptConfig.config) === null || _a === void 0 ? void 0 : _a.compilerOptions) {
            typeScriptConfig.config.compilerOptions.noUnusedLocals = false;
        }
        return typeScriptConfig;
    }
    static escapeRegExp(rawString) {
        return rawString.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    }
    async compileSnippets(documentationFiles) {
        try {
            await this.cleanWorkingDirectory();
            await fsExtra.ensureDir(this.workingDirectory);
            const examples = await this.extractAllCodeBlocks(documentationFiles);
            return await Promise.all(examples.map(async (example) => await this.testCodeCompilation(example)));
        }
        finally {
            await this.cleanWorkingDirectory();
        }
    }
    async cleanWorkingDirectory() {
        return await fsExtra.remove(this.workingDirectory);
    }
    async extractAllCodeBlocks(documentationFiles) {
        const importSubstituter = new LocalImportSubstituter_1.LocalImportSubstituter(this.packageDefinition);
        const codeBlocks = await Promise.all(documentationFiles.map(async (file) => await this.extractFileCodeBlocks(file, importSubstituter)));
        return codeBlocks.flat();
    }
    async extractFileCodeBlocks(file, importSubstituter) {
        const blocks = await CodeBlockExtractor_1.CodeBlockExtractor.extract(file);
        return blocks.map(({ code, type }, index) => {
            return {
                file,
                type,
                snippet: code,
                index: index + 1,
                sanitisedCode: this.sanitiseCodeBlock(importSubstituter, code),
            };
        });
    }
    sanitiseCodeBlock(importSubstituter, block) {
        const localisedBlock = importSubstituter.substituteLocalPackageImports(block);
        return localisedBlock;
    }
    async compile(code, type) {
        const id = process.hrtime.bigint().toString();
        const codeFile = path.join(this.workingDirectory, `block-${id}.${type}`);
        await fsExtra.writeFile(codeFile, code);
        const compiler = TSNode.create(this.compilerConfig);
        compiler.compile(code, codeFile);
    }
    removeTemporaryFilePaths(message, example) {
        const escapedCompiledDocsFolder = SnippetCompiler.escapeRegExp(path.basename(this.workingDirectory));
        const compiledDocsFilePrefixPattern = new RegExp(`${escapedCompiledDocsFolder}/block-\\d+\\.ts`, "g");
        return message.replace(compiledDocsFilePrefixPattern, (0, chalk_1.default) `{blue ${example.file}} → {cyan Code Block ${example.index}}`);
    }
    async testCodeCompilation(example) {
        try {
            await this.compile(example.sanitisedCode, example.type);
            return {
                snippet: example.snippet,
                file: example.file,
                index: example.index,
                linesWithErrors: [],
            };
        }
        catch (rawError) {
            const error = rawError instanceof Error ? rawError : new Error(String(rawError));
            error.message = this.removeTemporaryFilePaths(error.message, example);
            Object.entries(error).forEach(([key, value]) => {
                if (typeof value === "string") {
                    error[key] = this.removeTemporaryFilePaths(value, example);
                }
            });
            const linesWithErrors = new Set();
            if (error instanceof TSNode.TSError) {
                error.diagnostics.forEach((diagnostic) => {
                    const { start } = diagnostic;
                    if (typeof start === "undefined") {
                        return;
                    }
                    const lineNumber = [...example.sanitisedCode.substring(0, start)].filter((char) => char === "\n").length + 1;
                    linesWithErrors.add(lineNumber);
                });
            }
            return {
                snippet: example.snippet,
                error: error,
                linesWithErrors: [...linesWithErrors],
                file: example.file,
                index: example.index,
            };
        }
    }
}
exports.SnippetCompiler = SnippetCompiler;
